### PR TITLE
RavenDB-14248 Provide a way to export the raw Lucene index from RavenDB

### PR DESCRIPTION
**Pull request by ayende**

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 1]
```diff
@@ -3983,5 +3983,36 @@ public void AssertNotDisposed()
...
+            using (var tx = _environment.ReadTransaction())
+            {
+                var state = new Indexing.VoronState(tx);
+
+                var files = IndexPersistence.LuceneDirectory.ListAll(state);
+                totalNumberOfFiles = files.Length;
+                var buffer = new byte[64 * 1024];
+                foreach (var file in files)
+                {
+                    currentFile(file);
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/Indexes/Index.cs]: Can we have here a message like: Exporting file: {file}.

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 2]
```diff
@@ -3983,5 +3983,36 @@ public void AssertNotDisposed()
...
+                    using (var input = IndexPersistence.LuceneDirectory.OpenInput(file, state))
+                    using (var output = File.Create(Path.Combine(path, file)))
+                    {
+                        var len = input.Length(state);
+                        while (len > 0)
+                        {
+                            int read = (int)Math.Min(buffer.Length, len);
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/Indexes/Index.cs]: And progress here (every 500ms..)
Copied 10MB/500MB.

File: src/Raven.Server/Documents/Handlers/Admin/AdminIndexHandler.cs [hunk 3]
```diff
@@ -231,5 +235,105 @@ public Task Disable()
...
+            var operationId = Database.Operations.GetNextOperationId();
+            var token = CreateTimeLimitedQueryOperationToken();
+            var details = new DumpIndexOperationDetails
+            {
+                Index = name,
+            };
+
+            var task = Database.Operations.AddOperation(
+                Database,
+                "Dump index " + path,
```
Comment (lahma) [inline src/Raven.Server/Documents/Handlers/Admin/AdminIndexHandler.cs]: Was this supposed to be `"Dump index " + name`, or maybe even `"Dump index " + name + " to " + path` ?

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 1 above)
Comment (ayende) [reply src/Raven.Server/Documents/Indexes/Index.cs]: done

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 2 above)
Comment (ayende) [reply src/Raven.Server/Documents/Indexes/Index.cs]: done

File: src/Raven.Server/Documents/Handlers/Admin/AdminIndexHandler.cs (hunk 3 above)
Comment (ayende) [reply src/Raven.Server/Documents/Handlers/Admin/AdminIndexHandler.cs]: Yes, fixed

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 4]
```diff
@@ -3983,5 +3984,62 @@ public void AssertNotDisposed()
             if (_disposeOne.Disposed)
                 ThrowObjectDisposed();
         }
+
+        public int Dump(string path, Action<IOperationProgress> onProgress)
+        {
+            if (Directory.Exists(path) == false)
+                Directory.CreateDirectory(path);
+
+            using (var tx = _environment.ReadTransaction())
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/Index.cs:4001]: I'm wondering if we should not put `CurrentlyInUse` here?

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 5]
```diff
@@ -3983,5 +3984,62 @@ public void AssertNotDisposed()
...
+                    using (var output = File.Create(Path.Combine(path, file)))
+                    {
+                        var len = input.Length(state);
+                        string message = "Exporting file: " + file;
+                        onProgress(new AdminIndexHandler.DumpIndexProgress
+                        {
+                            Message = message,
+                            TotalFiles = files.Length,
+                            ProcessedFiles = currentFile,
+                            TotalBytes = len
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/Index.cs]: This doesn't look like TotalBytes? It is per file?

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 6]
```diff
@@ -3983,5 +3984,62 @@ public void AssertNotDisposed()
...
+                            output.Write(buffer, 0, read);
+                            if (sp.ElapsedMilliseconds > 1000)
+                            {
+                                onProgress(new AdminIndexHandler.DumpIndexProgress
+                                {
+                                    Message = message,
+                                    TotalFiles = files.Length,
+                                    ProcessedFiles = currentFile,
+                                    TotalBytes = len,
+                                    ProcessedBytes = totalRead
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/Index.cs]: this is also per file? Shouldn't this be global?

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 4 above)
Comment (ayende) [reply src/Raven.Server/Documents/Indexes/Index.cs:4001]: Yes, I think that would be good

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 5 above)
Comment (ayende) [reply src/Raven.Server/Documents/Indexes/Index.cs]: That is the total bytes for _this_ file, yes. Need to change this

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 4 above)
Comment (ayende) [reply src/Raven.Server/Documents/Indexes/Index.cs:4001]: done

**Review by ppekrol — APPROVED** [review]

**Comment by ml054** [comment]
I've created new PR here: https://github.com/ravendb/ravendb/pull/9881

as I don't have push rights to @ayende repo.

**Comment by arekpalinski** [comment]
Merged in https://github.com/ravendb/ravendb/pull/9881

**Pull request closed without merging**